### ISSUE TITLE
Regression in handling kwargs arguments

### ISSUE THREAD

**Issue by last-partizan** [microsoft/pyright#5545]
**Describe the bug**
I'm using `unittest.mock.patch` in my tests, and it has ability to pass kwargs to `configure_mock` later, so i'm using code like this:

```python
from unittest.mock import patch


@patch("urllib.request.urlopen", **{"return_value.status": 200})
def test1(*mocks):
    import urllib.request

    response = urllib.request.urlopen("https://localhost/")

    assert response.status == 200
```

Starting from `pyright 1.1.312` this code throws `error: Argument of type "int" cannot be assigned to parameter "create" of type "bool" in function "__call__"`

**Expected behavior**
It works fine.

**Code or Screenshots**

Here's minimal reproduction with plain function:

```python
from typing import Any


def func(create: bool = False, **kwargs: Any):
    ...


func(**{"k": 1})
```

```
> pyright minimal.py
/home/serg/src/pyright-examples/minimal.py
  /home/serg/src/pyright-examples/minimal.py:8:8 - error: Argument of type "int" cannot be assigned to parameter "create" of type "bool" in function "func"
    "int" is incompatible with "bool" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

**VS Code extension or command-line**
pyright version 1.1.318 (bug started appearing in 1.1.312)

**Additional context**

Mypy has the same issue.
```
> mypy minimal.py
minimal.py:8: error: Argument 1 to "func" has incompatible type "**Dict[str, int]"; expected "bool"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

**Comment by erictraut** [comment]
This change was intentional, so I don't consider it a bug.

Both behaviors (the one that pyright had before 1.1.312 and the one after) are defensible from a type checking perspective. Mypy and pyright previously had different behaviors, which was causing inconsistencies for overload matching. For that reason, I decided to switch pyright's behavior to match mypy's in this case.

Here are a couple of workarounds:

```python
# Specify that some of your function's parameters are positional-only
def func(create: bool = False, /, **kwargs: Any):
    ...
```

```python
# Specify the first argument explicitly
func(False, **{"k": 1})
```

**Comment by erictraut** [comment]
Other options include:
1. Don't use unpacked dictionaries to pass arguments. pass them by name instead. (`func(k=1)`)
2. Use a TypedDict to define the dict that holds the parameters that you're unpacking. When unpacking the TypedDict, a type checker will then know which attributes are present and the types of their associated values.
3. Disable type checking for your test modules.
4. Use `# type: ignore` or `# pyright: ignore` comments to suppress errors on specific lines.

**Comment by last-partizan** [comment]
> Behavior Change: Changed the matching of an unpacked dictionary passed as an argument to a callable when keyword arguments are present and have default values. Previously, the types of all such parameters were not checked against the unpacked dict type (under the assumption that the default argument would satisfy the parameter in this case). The new behavior matches that of mypy and assumes that the unpacked dict may supply arguments for all otherwise-unmatched keyword parameters even if they have default argument values.

I found this in changelog to 1.1.312 release, and i understand why it might be useful.

Have you considered making this an option instead of default behaviour? or maybe default, but with an option to disable it, something like `pyright: unpackedDictTypeChecking=false`. It would be extremely useful for writing tests.

And, i think i have seen an suggestion somewhere in the mypy issues, to infer in-place-unpacked dictionaries as typed dicts, or something like this. It would solve this issue as well, without disabling type checking or extra boilerplate.

For now i'd probably use workaround, like 
```python

def safeunpack(d) -> Any:
    return d

patch("module", **safeunpack({"args": ...}))
```

Anyway, thanks for maintaining this project, its super useful :)

**Comment by last-partizan** [comment]
I have found this discussion, and i like the idea of having `A_DICT: TypedDict = ...` style inference.

https://github.com/python/typing/discussions/1388#discussioncomment-5564921

Hope it gets traction and maybe somewhere in the future we can have unpacked dict inference as well.

**Comment by erictraut** [comment]
> Have you considered making this an option instead of default behaviour?

No, this is not the sort of thing we would want to make an option. The goal is for this behavior to be consistent across type checkers so library authors can rely on consistent behavior for overload matching, etc.
